### PR TITLE
Add MLT chain and link classes

### PR DESCRIPTION
**Pull request by bmatherly**
Also add the first link module: timeremap

### CODE REVIEW

File: src/framework/mlt.vers
Comment (ddennedy) [inline src/framework/mlt.vers:580]: MLT version 6.24.0 was released without this. So, we need to change this symbol version to the next release 6.26.0

File: src/mlt++/mlt++.vers
Comment (ddennedy) [inline src/mlt++/mlt++.vers:616]: Same thing here with regards to symbol version.

**Review by ddennedy — CHANGES_REQUESTED** [review]
I am not close to finish the review, but I wanted to get this feedback to you.

File: src/framework/mlt_producer.h
Comment (ddennedy) [inline src/framework/mlt_producer.h:83]: spelling error: "to s" instead of "to a"

File: src/framework/mlt_producer.h
Comment (ddennedy) [inline src/framework/mlt_producer.h:90]: Since this adds new members to a public struct, it does break ABI. That is OK for now but FYI. Hold off on the symbol version updates. Looks like this might be 7_0_0. In new major version, we can change some things to a private. But also keep in mind that you want to avoid breaking ABI a good while after upgrading it.

File: src/framework/mlt_producer.h
Comment (ddennedy) [inline src/framework/mlt_producer.h:86]: a `mlt_position` is not a "frame pointer by reference"

File: src/modules/avformat/factory.c
Comment (ddennedy) [inline src/modules/avformat/factory.c:434]: I was intentionally hiding service metadata for "avformat-novalidate" since it merely shadows "avformat" and needs special care to use. IOW, you cannot simply echo the avformat metadata. It could have its own abbreviated metadata that explains purpose and requirements and then says to see avformat. Do you need it for some reason to make this work?

File: src/modules/core/producer_melt.c
Comment (ddennedy) [inline src/modules/core/producer_melt.c:435]: The new `-chain` and `-link` options should be added to usage in `src/melt/melt.c` and `docs/melt.1`

File: src/modules/xml/producer_xml.c
Comment (ddennedy) [inline src/modules/xml/producer_xml.c:1607]: We need to update `src/modules/xml/mlt-xml.dtd` with the new `chain` and `link` elements.

File: src/framework/mlt_types.h
Comment (ddennedy) [inline src/framework/mlt_types.h:153]: Since we are breaking API, I want to change all these enum symbols to start like `mlt_type_...`

**Comment by ddennedy** [comment]
We probably need to add the new service types to `mlt_parser`.

File: src/modules/avformat/factory.c
Comment (bmatherly) [reply src/modules/avformat/factory.c:434]: A chain loads the metadata for encapsulated producer so that it can "reflect" the properties on itself. 
https://github.com/mltframework/mlt/blob/cb4a648fd4c688ee328cc53b61e7ca2d68cd22f4/src/framework/mlt_chain.c#L129
This avoids the need for the application to be aware that it is operating on a chain instead of directly on a producer.

If avformat-novalidate is the only exception, I could check for service "avformat-novalidate" and then load "avformat" in that specific case. Let me know if you prefer that.

**Comment by bmatherly** [comment]
> We probably need to add the new service types to `mlt_parser`.

I wondered about that. What is mlt_parser? And how do I test it?

**Comment by ddennedy** [comment]
> > We probably need to add the new service types to `mlt_parser`.
> 
> I wondered about that. What is mlt_parser? And how do I test it?

It traverses the graph. Shotcut uses it to find services in memory, sometime to update or run operations like replace.

File: src/framework/mlt.vers
Comment (bmatherly) [reply src/framework/mlt.vers]: OK. We will change it to 7.0.0 when the time comes.

File: src/framework/mlt_producer.h
Comment (bmatherly) [reply src/framework/mlt_producer.h]: Done

File: src/framework/mlt_producer.h
Comment (bmatherly) [reply src/framework/mlt_producer.h]: done

File: src/modules/core/producer_melt.c
Comment (bmatherly) [reply src/modules/core/producer_melt.c:435]: done

File: src/modules/xml/producer_xml.c
Comment (bmatherly) [reply src/modules/xml/producer_xml.c:1607]: done

File: src/framework/mlt_types.h
Comment (bmatherly) [reply src/framework/mlt_types.h]: done

**Comment by bmatherly** [comment]
> We probably need to add the new service types to mlt_parser.

Done

**Review by ddennedy — CHANGES_REQUESTED** [review]

File: src/modules/avformat/factory.c
Comment (ddennedy) [reply src/modules/avformat/factory.c]: "avformat-novalidate" is currently the only producer like this, but I could see this coming to qimage and pixbuf. How about handling anything ending with "-novalidate"?

File: src/framework/mlt.vers
Comment (ddennedy) [reply src/framework/mlt.vers]: I plan to merge this soon after the next small changes into a v7 development branch. So, let's go ahead and change the symbol version to MLT_7.0.0.

File: src/framework/mlt_link.h
Comment (ddennedy) [reply src/framework/mlt_link.h:63]: Forget about this now. I want to do this in a separate commit in the v7 branch. I do not want to make this PR more unwieldy.

File: src/framework/mlt.vers
Comment (bmatherly) [reply src/framework/mlt.vers]: done

File: src/modules/avformat/factory.c
Comment (bmatherly) [reply src/modules/avformat/factory.c]: done

File: src/framework/mlt_parser.h
Comment (bmatherly) [reply src/framework/mlt_parser.h]: done

File: src/mlt++/mlt++.vers
Comment (bmatherly) [reply src/mlt++/mlt++.vers]: This has been inconvenient for me in the past as well. I usually work by inspection and then I make sure that Shotcut will link against the resulting shared libraries. I searched for tools to check syntax/correctness of version scripts, but I did not find anything. If you have any idea of how we could automate checking the file, I volunteer to write a test for it.

File: configure
Comment (ddennedy) [inline configure:3]: should be 7.0.0

File: src/mlt++/mlt++.vers
Comment (ddennedy) [reply src/mlt++/mlt++.vers]: I found `readelf --dyn-syms -W ... ` in https://developers.redhat.com/blog/2019/08/01/how-the-gnu-c-library-handles-backward-compatibility/
And one can audit the output for anomalies, but I am not certain the C++ mangling helps to verify the correctness of signatures.

File: configure
Comment (bmatherly) [reply configure]: Done

File: src/mlt++/mlt++.vers
Comment (bmatherly) [reply src/mlt++/mlt++.vers]: readelf helps a little bit for inspection of the C library. But the mangling makes the output unusable for MLT++

**Comment by bmatherly** [comment]
By the way, whenever the time is right to merge this, feel free to squash everything if that keeps thing cleaner.

**Comment by ddennedy** [comment]
I merged this to a new v7 branch I pushed, but it does not show this as merged. I would rather not close this without showing it merged but can if it becomes necessary. Please try to rebase it on the v7 branch that is now in the repo.

**Comment by bmatherly** [comment]
I think this should be ready to go now.

**Pull request merged**
